`IndexSplit` for hierarchical parallelism (#981)
parthenon-hpc-lab/parthenon: doc/sphinx/src/nested_par_for.rst

``#pragma omp simd`` to vectorize the loop, which typically gives better
vectorization loops than ``PAR_LOOP_INNER_LAYOUT=TVR_INNER_LOOP`` on
CPUs and so is the default on CPUs.

Performance Considerations
---------------------------

Hierarchical parallelism can produce very performant code, but a
deeper awareness of how hardware is mapped to threads is required to
get optimal performance. Here we list a few strategies/considerations.

* On CPU, with `SIMDFOR_INNER_LOOP` you may have trouble vectorizing
  unless you help the compiler along. One way to do so is to work with
  raw pointers to contiguous memory, rather than working with views
  and strides. Even for stencil ops, if you can pull out pointers that
  represent the different points on the stencil, this can help with

>   vectorization.

Reviewer: This has been echoed by multiple talks at KUG

Author: Nice! It's good to hear others have had the same experience as us.

* Similarly on CPUs, due to the cost of starting up a vector op,
  vectorization will only be a performance win if there's enough work
  in the inner loop. A minimum of 16 points is required for the op to
  vectorize at all. Experience shows, however, that at least 64 is
  really required to see big wins. One strategy for providing enough
  vector cells in the inner loop is to do a 1D ``SIMDFOR`` inner loop
  but combine the ``j`` and ``i`` indices by simply looping over the
  contiguous memory in a rasterized plane on a block.

> * On Cuda GPUs, the outer loop typically maps to warps, while the
>   inner maps to threads. To see good performance, you must both
>   provide enough work in the inner loop to fill a warp and enough work
>   in the outer loop to fill all warps.

Reviewer: ```suggestion
* On Cuda GPUs, the outer loop typically maps to blocks, while the
  inner maps to threads. To see good performance, you must both
  provide enough work in the inner loop to create enough threads to fill a streaming multiprocessor (SM) with multiple warps to take advantage of pipelining and enough work
  in the outer loop to create enough blocks to fill all SMs on the GPU divided by the number of simultaneous streams. The number of warps in flight on the inner loop per SM (which is related to "occupancy") will depend positively on length of the inner loop and negatively on higher shared memory usage (or scratch pad memory in Kokkos parlance) and higher register usage.
```

I think this is more accurate to the hardware but might be too verbose. Has this been inline with your performance testing? I'm drawing from experience with optimizing GPU code but outside of Parthenon.

Reviewer: We can add the equivalent AMD lingo too. The performance considerations are largely the same, except that scratch pad is much smaller and wavefronts are twice as large as warps. (Warps=Wavefronts, Blocks=blocks, SM=CU, shared memory = LocalDataStore(LDS) ).

Author: Nice thanks for clarifying this. I added an additional sentence summarizing your comment on AMD GPUs as well.

Author: > Has this been inline with your performance testing?

Yeah you're spot on.

IndexSplit
-------------

To balance the CPU vs GPU hardware considerations of hierarchical
parallelism, ``Parthenon`` provides a utility, the ``IndexSplit``
class, defined in the ``utils/index_split.hpp`` header file and
available in ``<parthenon/package.hpp>`` in the

> ``parthenon::package::prelude`` namespace. The ``IndexSplit`` class

Reviewer: Can you give a high-level description of how `IndexSplit` behaves for CPUs and GPUs? How it might benefit performance? In what scenario should I use `IndexsSplit`?

Author: Yeah I added a paragraph on the context/benefit of `IndexSplit` and also some roughly guidelines for the optimal split on CPUs vs GPUs. Let me know what you think.

Reviewer: I'm not seeing this new paragraph in the docs. Has it been pushed up? Could you comment on it?

Author: Sorry I forgot to push yesterday. It's there now.

> * On Cuda GPUs, the outer loop typically maps to blocks, while the
>   inner maps to threads. To see good performance, you must both
>   provide enough work in the inner loop to create enough threads to fill a streaming multiprocessor (SM) with multiple warps to take advantage of pipelining and enough work
>   in the outer loop to create enough blocks to fill all SMs on the GPU divided by the number of simultaneous streams. The number of warps in flight on the inner loop per SM (which is related to "occupancy") will depend positively on length of the inner loop and negatively on higher shared memory usage (or scratch pad memory in Kokkos parlance) and higher register usage.

Reviewer: ```suggestion
* On GPUs, the outer loop typically maps to blocks, while the
  inner maps to threads. To see good performance, you must both
  provide enough work in the inner loop to create enough threads to fill in CUDA terms a streaming multiprocessor (SM, equivalent to a Compute Unit or CU on AMD GPUs) with multiple warps (or wavefronts for AMD) to take advantage of pipelining and enough work
  in the outer loop to create enough blocks to fill all SMs on the GPU divided by the number of simultaneous streams. The number of warps in flight on the inner loop per SM (which is related to "occupancy") will depend positively on length of the inner loop and negatively on higher shared memory usage (scratch pad memory in Kokkos parlance and Local Data Store or LDS on AMD GPUs) and higher register usage. Note that the number of SMs and the available shared memory and registers per SM will vary between GPU architectures and especially between GPU vendors.
```

Added a bit of translation for AMD GPUs. I doubt many will care, but it's there now

Author: Nice. Thanks!

can be constructed as

.. code:: cpp

  IndexSplit(MeshData<Real> md, IndexDomain domain, const int nkp, const int njp);

where here ``md`` is ``MeshData`` object on which you want to

> operatore. ``domain`` specifies where in the ``MeshBlock`` you wish to

Reviewer: ```suggestion
where here ``md`` is a ``MeshData`` object on which you want to
operate. ``domain`` specifies where in the ``MeshBlock`` you wish to
```

operate, for example ``IndexDomain::Interior``. ``nkp`` and ``njp``
are the number of points in ``X3`` and ``X2`` respectively that are in

> the outer loop. All remaining points are in the inner loop. Typically

Reviewer: So  unless both are `all_outer` or `nkp` equals the full index range, then some fraction of the k iteration or j iteration will iterate within the same team, right?

Reviewer: The parameters for best performance would seem to vary per kernel and GPU hardware, I would think. Can we provide functionality for best guesses that depend on architecture? Like `Kokkos::HintLightWeight`?

Might be beyond the scope of this PR.

Reviewer: ```suggestion
operate, for example ``IndexDomain::Interior``. ``nkp`` and ``njp``
are the number of points in ``X3`` and ``X2`` respectively that are in
the outer loop. All remaining points are in the inner loop; each team will iterate over multiple `k` and/or `j` indices to cover the specified `k/j` range. Typically
```

Author: > So unless both are all_outer or nkp equals the full index range, then some fraction of the k iteration or j iteration will iterate within the same team, right?

That's right

> The parameters for best performance would seem to vary per kernel and GPU hardware, I would think. Can we provide functionality for best guesses that depend on architecture? Like Kokkos::HintLightWeight?
> 
> Might be beyond the scope of this PR.

I like this idea, but I do think this is outside of the scope of this one.

  IndexSplit idx_sp(md, IndexDomain::interior, nkp, njp);
  
  // Request maximum size in i and j in the inner loop, for scratch
  const int Ni = idx_sp.get_max_ni();
  const int Nj = idx_sp = get_max_nj();
  const in tNmax = Ni * Nj;
  
  // single scratch array for i,j
  auto scratch_size = ScratchPad1D<Real>::shmem_size(Nmax);
  constexpr int scratch_level = 0;
  
  // Par for
  par_for_outer(
	  DEFAULT_OUTER_LOOP_PATTERN, "KernalOuter", DevExecSpace(), scratch_size,
	  scratch_level, 0, nblocks - 1, 0, idx_sp.outer_size() - 1,
	  KOKKOS_LAMBDA(team_mbr_t member, const int b, const int outer_idx) {
	    ScratchPad1D<Real> scratch(member.team_scratch(scratch_level), Nmax);
	    // Get index ranges. Note they depend on where we are in the outer index!
	    // These give us a sense for where we are in k,j space
	    const auto krange = idx_sp.GetBoundsK(outer_idx);
	    const auto jrange = idx_sp.GetBoundsJ(outer_idx);
	    // This is the loop of contiguous inner memory. May contain i and j!
	    const auto irange = idx_sp.GetInnerBounds(jrange);

	    // Whatever part of k is not in the outer loop can be looped over
	    // with a normal for loop here

> 	    for (int k = krange.s; k <= krange.e; ++k) {

Reviewer: Since we're performing the same iteration for each thread in the team, this isn't improving occupancy by increasing number of threads per block but increasing work per thread, correct?

Author: Yes---that's right.

  number of simultaneous streams. The number of warps in flight on the
  inner loop per SM (which is related to "occupancy") will depend
  positively on length of the inner loop and negatively on higher
  shared memory usage (scratch pad memory in Kokkos parlance and Local
  Data Store or LDS on AMD GPUs) and higher register usage. Note that
  the number of SMs and the available shared memory and registers per
  SM will vary between GPU architectures and especially between GPU
  vendors.

IndexSplit
-------------

To balance the CPU vs GPU hardware considerations of hierarchical
parallelism, ``Parthenon`` provides a utility, the ``IndexSplit``
class, defined in the ``utils/index_split.hpp`` header file and
available in ``<parthenon/package.hpp>`` in the
``parthenon::package::prelude`` namespace.

In our experience ``IndexSplit`` is most beneficial when working with
small meshblocks, especially in two dimensions. For small blocks, we
want vectorized operations over contiguous memory for our innermost
loop, but we want that loop to contain enough work for, e.g., vector
ops to function. We have often found that the optimal split is to fuse

> j, and i into the inner loop and use k and blocks in the outer loop.

Author: @forrestglines here's the new paragraph

Reviewer: I see! So does this give better performance on GPUs and on CPUs? The mention of vectorization suggests that this is a CPU improvement. In my mental model of GPUs launching fewer blocks vs. many blocks here could only have a detrimental performance if `k` is less than the number of SMs. How did this affect GPU performance in your experience?

Reviewer: Could this functionality be inserted into `par_for` so that all local operations use this more optimal arrangement of teams?

Author: > I see! So does this give better performance on GPUs and on CPUs? The mention of vectorization suggests that this is a CPU improvement. 

Yeah this is a CPU optimization. I can add a comment saying so. I'm not sure what the optimal choice is for GPUs yet. We haven't explored it too much.

> Could this functionality be inserted into par_for so that all local operations use this more optimal arrangement of teams?

Possibly... I'd say we should put that in in a separate PR though. :)

  in the inner loop. A minimum of 16 points is required for the op to
  vectorize at all. Experience shows, however, that at least 64 is
  really required to see big wins. One strategy for providing enough
  vector cells in the inner loop is to do a 1D ``SIMDFOR`` inner loop
  but combine the ``j`` and ``i`` indices by simply looping over the
  contiguous memory in a rasterized plane on a block.
* On GPUs, the outer loop typically maps to blocks, while the inner
  maps to threads. To see good performance, you must both provide
  enough work in the inner loop to create enough threads to fill in
  CUDA terms a streaming multiprocessor (SM, equivalent to a Compute
  Unit or CU on AMD GPUs) with multiple warps (or wavefronts for AMD)
  to take advantage of pipelining and enough work in the outer loop to
  create enough blocks to fill all SMs on the GPU divided by the
  number of simultaneous streams. The number of warps in flight on the
  inner loop per SM (which is related to "occupancy") will depend
  positively on length of the inner loop and negatively on higher
  shared memory usage (scratch pad memory in Kokkos parlance and Local

>   Data Store or LDS on AMD GPUs) and higher register usage. Note that

Reviewer: ```suggestion
  Data Share or LDS on AMD GPUs) and higher register usage. Note that
```
My typo

blocks, we want vectorized operations over contiguous memory for our
innermost loop, but we want that loop to contain enough work for,
e.g., vector ops to function. We have often found that the optimal
split is to fuse j, and i into the inner loop and use k and blocks in
the outer loop.

The ``IndexSplit`` class can be constructed as

.. code:: cpp

  IndexSplit(MeshData<Real> md, IndexDomain domain, const int nkp, const int njp);

where here ``md`` is a ``MeshData`` object on which you want to
operate. ``domain`` specifies where in the ``MeshBlock`` you wish to
operate, for example ``IndexDomain::Interior``. ``nkp`` and ``njp``
are the number of points in ``X3`` and ``X2`` respectively that are in
the outer loop. All remaining points are in the inner loop; each team
will iterate over multiple `k` and/or `j` indices to cover the
specified `k/j` range. Typically ``MeshBlock`` index in the pack is
also assumed to be in the outer loop. ``nkp`` and ``njp`` also accept
special flags ``IndexSplit::all_outer`` and ``IndexSplit::no_outer``,
which specify that all and none of the indices in that direction
should be in the outer loop.

> 

Reviewer 2: Maybe comment somewhere in here that the nkp points in the k-direction are not included in the innermost loop bounds (in contrast to the njp points).

Reviewer 2: Since this is the case, I assume I can choose something like `nkp = 8` and `njp = 1` and everything will still work correctly?

Author: > Maybe comment somewhere in here that the nkp points in the k-direction are not included in the innermost loop bounds (in contrast to the njp points).

Done

> Since this is the case, I assume I can choose something like nkp = 8 and njp = 1 and everything will still work correctly?

Yep.

  using parthenon::IndexSplit;
  using parthenon::par_for_outer;
  using parthenon::par_for_inner;
  using parthenon::team_mbr_t;
  // Initialize index split object
  IndexSplit idx_sp(md, IndexDomain::interior, nkp, njp);
  
  // Request maximum size in i and j in the inner loop, for scratch
  const int Ni = idx_sp.get_max_ni();
  const int Nj = idx_sp = get_max_nj();
  const in tNmax = Ni * Nj;
  
  // single scratch array for i,j
  auto scratch_size = ScratchPad1D<Real>::shmem_size(Nmax);
  constexpr int scratch_level = 0;
  
  // Par for
  par_for_outer(
	  DEFAULT_OUTER_LOOP_PATTERN, "KernalOuter", DevExecSpace(), scratch_size,
	  scratch_level, 0, nblocks - 1, 0, idx_sp.outer_size() - 1,
	  KOKKOS_LAMBDA(team_mbr_t member, const int b, const int outer_idx) {
	    ScratchPad1D<Real> scratch(member.team_scratch(scratch_level), Nmax);
	    // Get index ranges. Note they depend on where we are in the outer index!
	    // These give us a sense for where we are in k,j space
	    const auto krange = idx_sp.GetBoundsK(outer_idx);
	    const auto jrange = idx_sp.GetBoundsJ(outer_idx);
	    // This is the loop of contiguous inner memory. May contain i and j!

> 	    const auto irange = idx_sp.GetInnerBounds(jrange);

Reviewer 2: Maybe call this `flattened_inner_ijrange` instead of `irange` for maximum clarity?

Author: done

	    // Whatever part of k is not in the outer loop can be looped over
	    // with a normal for loop here
	    for (int k = krange.s; k <= krange.e; ++k) {

	      // pull out a pointer some variable in some pack. Note
	      // we pick the 0th index of i at k and jrange.s

> 	      Real *var = &pack(b, ivar, k, jrange.s, 0);

Reviewer 2: ```suggestion
	      Real *var = &pack(b, ivar, k, jrange.s, irange.s);
```
